[AMDGPU] Infer amdgpu-no-flat-scratch-init attribute in AMDGPUAttributor

jwanggit86 opened the pull request:
The AMDGPUAnnotateKernelFeatures pass infers the "amdgpu-calls" and "amdgpu-stack-objects" attributes, which are used to infer whether we need to initialize flat scratch. This is, however, not precise. Instead, we should use AMDGPUAttributor and infer amdgpu-no-flat-scratch-init on kernels. Refer to https://github.com/llvm/llvm-project/issues/63586 .

jwanggit86 committed:
[AMDGPU] Infer amdgpu-no-flat-scratch-init attribute in AMDGPUAttributor

The AMDGPUAnnotateKernelFeatures pass infers the "amdgpu-calls" and
"amdgpu-stack-objects" attributes, which are used to infer whether we need to
initialize flat scratch. This is, however, not precise. Instead, we should use
AMDGPUAttributor and infer amdgpu-no-flat-scratch-init on kernels.
Refer to https://github.com/llvm/llvm-project/issues/63586 .

jdoerfert commented on llvm/lib/Target/AMDGPU/AMDGPUAttributor.cpp:
```diff
@@ -433,6 +433,19 @@ struct AAAMDAttributesFunction : public AAAMDAttributes {
       return;
     }
+
+    bool HasAllocaOrASCast = false;
+    for (BasicBlock &BB : *F) {
+      for (Instruction &I : BB) {
+        if (isa<AllocaInst>(I) || isa<AddrSpaceCastInst>(I)) {
+          HasAllocaOrASCast = true;
+          removeAssumedBits(FLAT_SCRATCH_INIT);
+          break;
+        }
+      }
+      if (HasAllocaOrASCast)
+        break;
+    }
```
At some point, you need to look at the non-call instructions.
It makes sense to do that in update *if* we ever do value simplification or dead code elimination as part of the pass. Otherwise, init is fine, I think.

jdoerfert commented on llvm/lib/Target/AMDGPU/AMDGPUAttributor.cpp:
```diff
@@ -433,6 +433,19 @@ struct AAAMDAttributesFunction : public AAAMDAttributes {
     }
+
+    bool HasAllocaOrASCast = false;
+    for (BasicBlock &BB : *F) {
```
isn't there a `instructions(F)`? Then you don't need 2 breaks, no bool, much easier.

jwanggit86 replied:
Done.

jdoerfert commented on llvm/lib/Target/AMDGPU/AMDGPUAttributor.cpp:
```diff
@@ -677,6 +693,33 @@ struct AAAMDAttributesFunction : public AAAMDAttributes {
+      const Value *CalleeOp = CB.getCalledOperand();
+      const Function *Callee = dyn_cast<Function>(CalleeOp);
+      if (!Callee) // indirect call
+        return CB.isInlineAsm();
```
You should replace all this handling, starting from here till the end of this lambda, with a call to 
```
 /// Check \p Pred on all potential Callees of \p CB.
  ///
  /// This method will evaluate \p Pred with all potential callees of \p CB as
  /// input and return true if \p Pred does. If some callees might be unknown
  /// this function will return false.
  bool checkForAllCallees(
      function_ref<bool(ArrayRef<const Function *> Callees)> Pred,
      const AbstractAttribute &QueryingAA, const CallBase &CB);
```
This will work for indirect calls as well.

jwanggit86 replied:
I'm not sure I follow. This `checkForAllCallees()` function requires a pred. So we wrap the existing code of the lambda in a new lambda, and pass the new lambda to `checkFroAllCallees()`? Also, could you pls explain a little what the potential problem is without using this function?

jwanggit86 replied:
@jdoerfert ping here.

arsenm replied:
Instead of checking for the call instructions with checkForAllCallLikeInstructions, you instead use checkForAllCallees for all direct and possible indirect call candidates (not 100% sure how this interacts with asm). That way you don't have to concern yourself with indirect calls, since any possible call targets should also appear in the callees

jwanggit86 replied:
`checkForAllCallees()` requires an arg which is a `CallBase`, i.e., a call-like instruction. So, to use it, you need to get the call instruction first. On the other hand, `checkForAllCallLikeInstructions()` can give you all the call instructions. It's not clear to me how `checkForAllCallees` can replace `checkForAllCallLikeInstructions`.

ssahasra replied:
`checkForAllCallees()` seems fairly new. If I understand it correctly, we will need to move most of this lambda to a call-site attribute as a helper to do all the real work. Then `needFlatScratchInit()` will retrieve that attribute on every call-like instruction. Is that correct? It definitely looks cleaner that way, I think. But then, the same could be done for all the other function attributes being tracked here? Should that be captured as a pattern in the attributor itself? "a function attribute that is computed by examining the same attribute for every callsite".

jwanggit86 replied:
@ssahasra I fully agree with your description of how it works if `checkForAllCallees()` is used. Based on this, I'm not sure I see the advantage of using `checkForAllCallees()`. I think you still need to check all call-like instructions, and all it does is to move the real work (check for intrinsics, indirect calls etc) to this CallSite attribute. Furthermore, don't you need both a CallSite attribute and a Function attribute, because when you set up the CallSite attribute you need to retrieve some attribute from the callee? Besides, I want to mention that I was not able to find any use of `checkForAllCallees()` or any definition of CallSite attributes in the existing code.
@arsenm @jdoerfert Your thoughts?

arsenm replied:
This is backwards. checkForAllCallees's callback takes only the candidate callee function, not the callsite. The point is it passes possible indirect callees you cannot trivially infer from just the callsite. This does mean you can miss manually specified callsite attributes, but that rarely matters.

Just use checkForAllCallees.

ssahasra replied:
> This is backwards. checkForAllCallees's callback takes only the candidate callee function, not the callsite.

Ummm, no it does not? Unless we are looking at different things, it takes a `CallBase`:

```
   1737 bool Attributor::checkForAllCallees(
   1738     function_ref<bool(ArrayRef<const Function *>)> Pred,
   1739     const AbstractAttribute &QueryingAA, const CallBase &CB) {
```

> The point is it passes possible indirect callees you cannot trivially infer from just the callsite. This does mean you can miss manually specified callsite attributes, but that rarely matters.

Yes, that's the advantage.
 
> Just use checkForAllCallees.

But how exactly? When computing a function attribute, we have to do `getAAFor()` on every callsite in that function, right?

ssahasra replied:
> I was not able to find any use of `checkForAllCallees()` or any definition of CallSite attributes in the existing code. 

It is used through the `AACalleeToCallSite` helper class. This class is used by a lot of attributes that are computed over the callgraph. Take a look at how `AANoFree` is constructed for example. It is declared as both a function attribute and callsite attribute.

@arsenm @jdoerfert Your thoughts?

jwanggit86 replied:
If we were to use `callForAllCallees()`, based on my understanding, the code would look something like this:
```diff 
   bool needFlatScratchInit(Attributor &A) {
     // This is called on each callee; false means callee shouldn't have
     // no-flat-scratch-init.
     auto CheckForNoFlatScratchInit = [&](Instruction &I) {
       const auto &CB = cast<CallBase>(I);
-      const Value *CalleeOp = CB.getCalledOperand();
-      const Function *Callee = dyn_cast<Function>(CalleeOp);
-      if (!Callee) // indirect call
-        return CB.isInlineAsm();
-
-      if (Callee->isIntrinsic())
-        return Callee->getIntrinsicID() != Intrinsic::amdgcn_addrspacecast_nonnull;
-
-      const auto *CalleeInfo = A.getAAFor<AAAMDAttributes>(
-          *this, IRPosition::function(*Callee), DepClassTy::REQUIRED);
-      return CalleeInfo && CalleeInfo->isAssumed(FLAT_SCRATCH_INIT);
+      auto PredOnCallees = [&](ArrayRef<const Function *> Callees) {
+        bool Ret = true;
+        for (const Function *Callee : Callees) {
+          if (!Callee) { // indirect call
+            // non-asm indirect call is already handled in updateImpl()
+            assert(CB.isInlineAsm());
+            continue;
+          }
+
+          if (Callee->isIntrinsic()) {
+            Ret &= (Callee->getIntrinsicID() != Intrinsic::amdgcn_addrspacecast_nonnull);
+            if (!Ret)
+              return false;
+            continue;
+          }
+
+          const auto *CalleeInfo = A.getAAFor<AAAMDAttributes>(
+              *this, IRPosition::function(*Callee), DepClassTy::REQUIRED);
+          Ret &= (CalleeInfo && CalleeInfo->isAssumed(FLAT_SCRATCH_INIT));
+          if (!Ret)
+            return false;
+        }
+        return Ret;
+      };
+
+      return A.checkForAllCallees(PredOnCallees, *this, CB);
     };

     bool UsedAssumedInformation = false;
    // If any callee is false (i.e. need FlatScratchInit),
    // checkForAllCallLikeInstructions returns false, in which case this
    // function returns true.
    return !A.checkForAllCallLikeInstructions(CheckForNoFlatScratchInit, *this,
                                              UsedAssumedInformation);
  }
```
Here, in the lambda `CheckForNoFlatScratchInit`, a new lambda is created (`PredOnCallees`), and it simply calls `checkForAllCallees()` passing the new lambda. 
@arsenm @jdoerfert Pls let me know if this is what you had in mind.

arsenm replied:
I assume you shouldn't need to handle a null callee in the callback. Also, I don't think you have the CallBase here.

It's easier to review an actual patch, so submit a new one using checkForAllCallees?

jwanggit86 replied:
Code has been updated after fixing a bug regarding indirect calls of known callees. I'm not sure using `checkForAllCallees()` is necessary for 2 reasons: (1) code is now very simple, boiling down to just checking for intrinsics (2) the first part of `AAAMDAttributesFunction::updateImpl()` already does similar work as `checkForAllCallees()`:
```cpp
  ChangeStatus updateImpl(Attributor &A) override {
    ...
    // Check for Intrinsics and propagate attributes.
    const AACallEdges *AAEdges = A.getAAFor<AACallEdges>(
        *this, this->getIRPosition(), DepClassTy::REQUIRED);
    if (!AAEdges || AAEdges->hasNonAsmUnknownCallee())
      return indicatePessimisticFixpoint();
...
    for (Function *Callee : AAEdges->getOptimisticEdges()) { // Jun - what checkForAllCallees() does is essentially calling getOptimisticEdges() and then running the pred on the edges
      Intrinsic::ID IID = Callee->getIntrinsicID();
      if (IID == Intrinsic::not_intrinsic) {
        const AAAMDAttributes *AAAMD = A.getAAFor<AAAMDAttributes>(
            *this, IRPosition::function(*Callee), DepClassTy::REQUIRED);
        if (!AAAMD)
          return indicatePessimisticFixpoint();
        *this &= *AAAMD;
        continue;
      }
...
```

jwanggit86 replied:
On the other hand, a patch where `checkForAllCallees()` is used is as follows:
```diff
   bool needFlatScratchInit(Attributor &A) {
     assert(isAssumed(FLAT_SCRATCH_INIT)); // only called if the bit is still set
     auto CheckForNoFlatScratchInit = [&](Instruction &I) {
       const auto &CB = cast<CallBase>(I);
-      const Function *Callee = CB.getCalledFunction();
-
-      if (Callee && Callee->isIntrinsic())
-        return Callee->getIntrinsicID() !=
-               Intrinsic::amdgcn_addrspacecast_nonnull;
-
+      auto PredOnCallees = [&](ArrayRef<const Function *> Callees) {
+        for (const Function *Callee : Callees) {
+          if (Callee->isIntrinsic())
+            if (Callee->getIntrinsicID() == Intrinsic::amdgcn_addrspacecast_nonnull)
+              return false;
+        }
+        return true;
+      };
+      if(!A.checkForAllCallees(PredOnCallees, *this, CB))
+        return CB.isInlineAsm();
       return true;
     };

     bool UsedAssumedInformation = false;
     return !A.checkForAllCallLikeInstructions(CheckForNoFlatScratchInit, *this,
```

arsenm replied:
The point of using chekcForAllCallees is it also works for indirect calls, not that it makes the code simpler or better. You can do that in a follow up patch if you really want, but you should still use it

jwanggit86 commented on llvm/lib/Target/AMDGPU/AMDGPUAttributor.cpp:
```diff
@@ -677,6 +693,33 @@ struct AAAMDAttributesFunction : public AAAMDAttributes {
+    // no-flat-scratch-init.
+    auto CheckForNoFlatScratchInit = [&](Instruction &I) {
+      const auto &CB = cast<CallBase>(I);
+      const Value *CalleeOp = CB.getCalledOperand();
+      const Function *Callee = dyn_cast<Function>(CalleeOp);
```
There's a subtle difference between using `getCalledFuunction()` and using `getCalledOperand()`. In the following example,
```
define i32 @use_dispatch_ptr_ret_type() #1 {
  %dispatch.ptr = call ptr addrspace(4) @llvm.amdgcn.dispatch.ptr()
  store volatile ptr addrspace(4) %dispatch.ptr, ptr addrspace(1) undef
  ret i32 0
}

define float @func_indirect_use_dispatch_ptr_constexpr_cast_func() #1 {
  %f = call float @use_dispatch_ptr_ret_type()
  %fadd = fadd float %f, 1.0
  ret float %fadd
}
```
Note that callee's return type is i32 but caller casts it to float. Because of this, `getCalledFunction()` would return nullptr and eventually lead to incorrect analysis result.

arsenm replied:
Yes, it returns null when the call type mismatches the callee type. This is UB and doesn't matter; you don't need optimal handling. In the real world the call would have been replaced with unreachable

jwanggit86 replied:
@arsenm So you are saying that we should treat such cases as if the call didn't exist?

arsenm replied:
They should be treated like indirect calls

jwanggit86 commented on llvm/lib/Target/AMDGPU/AMDGPUAttributor.cpp:
```diff
@@ -677,6 +693,33 @@ struct AAAMDAttributesFunction : public AAAMDAttributes {
+        return CB.isInlineAsm();
+
+      if (Callee->isIntrinsic())
+        return true;
```
Did you mean:
```diff
       if (Callee->isIntrinsic())
-        return true;
+        return Callee->getIntrinsicID() != Intrinsic::amdgcn_addrspacecast_nonnull;
```

arsenm replied:
Something like that, in your visit function you need to treat the intrinsic exactly like a real addrspacecast instruction

ssahasra commented on llvm/lib/Target/AMDGPU/AMDGPUAttributor.cpp:
```diff
@@ -677,6 +693,33 @@ struct AAAMDAttributesFunction : public AAAMDAttributes {
+    bool UsedAssumedInformation = false;
+    // If any callee is false (i.e. need FlatScratchInit),
+    // checkForAllCallLikeInstructions returns false
+    return !A.checkForAllCallLikeInstructions(CheckForNoFlatScratchInit, *this,
```
Doing it like QueuePtr is certainly more "canonical'. Seems more preferable than doing it in `initialize()`, although the difference won't be noticeable unless the attributor is also simplifying the program at the same time.

jwanggit86 committed:
(1) Use getCalledFunction instead of getCalledOperand (2) other
minor code change based on reviews (3) fix test files.

jwanggit86 committed:
Previous code was incorrect for indirect calls of known callees.

jwanggit86 commented on llvm/test/CodeGen/AMDGPU/attributor-flatscratchinit.ll:
```diff
@@ -0,0 +1,914 @@
+;
+; GFX10-LABEL: define amdgpu_kernel void @with_region_to_flat_addrspacecast_cc_kernel(ptr addrspace(2) %ptr)
+; GFX10-SAME:  #[[ATTR_GFX10_NO_NOFSI2:[0-9]+]]
+  %stof = addrspacecast ptr addrspace(2) %ptr to ptr
```
Done.

jwanggit86 commented on llvm/lib/Target/AMDGPU/AMDGPUSubtarget.cpp:
```diff
@@ -1040,12 +1040,8 @@ GCNUserSGPRUsageInfo::GCNUserSGPRUsageInfo(const Function &F,
-  // to look for allocas.
-  const bool HasStackObjects = F.hasFnAttribute("amdgpu-stack-objects");
+  const bool NoFlatScratchInit =
+      F.hasFnAttribute("amdgpu-no-flat-scratch-init");
```
Done.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUAttributor.cpp:
```diff
@@ -433,6 +433,13 @@ struct AAAMDAttributesFunction : public AAAMDAttributes {
     }
+
+    for (Instruction &I : instructions(F)) {
+      if (isa<AllocaInst>(I) || isa<AddrSpaceCastInst>(I)) {
```
only addrspacecasts from address space 5 require flat_scr. Just an alloca does not imply flat_scr is required, only if the alloca is ever cast to flat. 


I suppose for robustness you could double check the alloca address space is correct

arsenm replied:
Approximately yes, but you should use dyn_cast and never static_cast

jwanggit86 replied:
> Approximately yes, but you should use dyn_cast and never static_cast

Doesn't the check `isa<AddrSpaceCastInst>(I)` already make sure that the instruction is a `AddrSpaceCastInst`?

jwanggit86 replied:
@arsenm ping.

ssahasra replied:
``static_cast`` is a C++ feature, but rarely used in LLVM code. Instead we use ``dyn_cast`` or ``cast``. You should be doing something like this:

`if (auto *AddrCastInst = dyn_cast<AddrSpaceCastInst>(&I)) { ... }`

jwanggit86 replied:
Thanks. This has been updated.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUAttributor.cpp:
```diff
@@ -677,6 +687,36 @@ struct AAAMDAttributesFunction : public AAAMDAttributes {
+      const auto &CB = cast<CallBase>(I);
+      const Function *Callee = CB.getCalledFunction();
+
+      if (Callee && Callee->isIntrinsic())
```
Don't need to check isIntrinsic, getIntrinsicID will just be not_intrinsic for an arbitrary call

jwanggit86 replied:
Good point.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUAttributor.cpp:
```diff
@@ -677,6 +687,36 @@ struct AAAMDAttributesFunction : public AAAMDAttributes {
+      // call, and (3)indirect call with known callees. For (2) and (3)
+      // updateImpl() already checked the callees and we know their
+      // FLAT_SCRATCH_INIT bit is set.
+      return true;
```
Can just return the boolean expression above

jwanggit86 replied:
Improved a little.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUSubtarget.cpp:
```diff
@@ -1054,12 +1054,6 @@ GCNUserSGPRUsageInfo::GCNUserSGPRUsageInfo(const Function &F,
   const CallingConv::ID CC = F.getCallingConv();
   const bool IsKernel =
       CC == CallingConv::AMDGPU_KERNEL || CC == CallingConv::SPIR_KERNEL;
-  // FIXME: Should have analysis or something rather than attribute to detect
-  // calls.
-  const bool HasCalls = F.hasFnAttribute("amdgpu-calls");
-  // FIXME: This attribute is a hack, we just need an analysis on the function
-  // to look for allocas.
-  const bool HasStackObjects = F.hasFnAttribute("amdgpu-stack-objects");
```
In the interest of making the test changes smaller, I think this part should be done in a separate patch (which also removes AMDGPUAnnotateKernelFeatures at the same time). This patch should just do the pure inference.

Although now that the attributor is moved, I'm worried about passes between amdgpuattributor and codegen introducing new addrspacecasts

jwanggit86 replied:
This is done now. Changes to AMDGPUAnnotateKernelFeatures will be in a separate PR.

jwanggit86 committed:
Undo changes in AMDGPUSubtarget.cpp to reduce impact on test files.
Those code changes will be in a follow-up PR.

jwanggit86 committed:
Fix tests after merge from main branch

jwanggit86 committed:
Undo changes made to AMDGPUSubtarget.cpp to limit impact on tests.
This undo is simply achieved by merging code from upstream because
a recent commit has changed that file.
The changes therein will be included in a separate PR.

jwanggit86 committed:
Code improvement in the lambda.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUAttributor.cpp:
```diff
@@ -678,6 +690,37 @@ struct AAAMDAttributesFunction : public AAAMDAttributes {
+
+    // This is called on each callee; false means callee shouldn't have
+    // no-flat-scratch-init.
+    auto CheckForNoFlatScratchInit = [&](Instruction &I) {
+      const auto &CB = cast<CallBase>(I);
```
I would hope FroAllCallLikeInstructions would have a CallBase typed argument to begin with

jwanggit86 commented on llvm/lib/Target/AMDGPU/AMDGPUAttributor.cpp:
```diff
@@ -434,6 +434,15 @@ struct AAAMDAttributesFunction : public AAAMDAttributes {
     }
+
+    for (Instruction &I : instructions(F)) {
+      if (isa<AddrSpaceCastInst>(I) &&
```
@arsenm Could you please give me an example of a constant expression having an addrSpaceCast?

jwanggit86 replied:
@arsenm Following constants to see if they contain addrSpaceCast is now done. An example is: `store i32 7, ptr addrspace(3) addrspacecast (ptr addrspace(5) null to ptr addrspace(3))`. 
However, I'm not sure it's required or even correct. For the above example, opt with -O2 would optimize away the addrspacecast, and the result would be the opposite.

arsenm replied:
5->3 is an illegal address space cast, but the round trip cast can fold away. You don't want the cast back to the original address space.

jwanggit86 replied:
This test has been added.

jwanggit86 committed:
(1) check constant for addrspacecast (2) remove alloca related tests.

jwanggit86 commented on llvm/lib/Target/AMDGPU/AMDGPUAttributor.cpp:
```diff
@@ -678,6 +690,37 @@ struct AAAMDAttributesFunction : public AAAMDAttributes {
+      // already removed in updateImpl() and execution won't reach here.
+      if (!Callee)
+        return true;
+      else
```
Done.

jwanggit86 committed:
Add a testcase for addrspacecast inside a constant.

jwanggit86 committed:
Update tests after merging from main.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUAttributor.cpp:
```diff
@@ -683,6 +706,59 @@ struct AAAMDAttributesFunction : public AAAMDAttributes {
+                                              UsedAssumedInformation);
+  }
+
+  bool constHasASCast(const Constant *C,
+                      SmallPtrSetImpl<const Constant *> &Visited) {
+    if (!Visited.insert(C).second)
+      return false;
+
+    if (const auto *CE = dyn_cast<ConstantExpr>(C))
+      if (CE->getOpcode() == Instruction::AddrSpaceCast &&
+          CE->getOperand(0)->getType()->getPointerAddressSpace() ==
+              AMDGPUAS::PRIVATE_ADDRESS)
+        return true;
+
+    for (const Use &U : C->operands()) {
+      const auto *OpC = dyn_cast<Constant>(U);
+      if (!OpC || !Visited.insert(OpC).second)
+        continue;
+
+      if (constHasASCast(OpC, Visited))
+        return true;
+    }
+    return false;
+  }
```
I do not want to duplicate the same function that already exists for the LDS case. Unify these.

We also should try to avoid doing this walk over all instructions through all constant expressions twice for the two attributes

jwanggit86 replied:
@arsenm What is the LDS case? LDS_KERNEL_ID? Are you saying there's another attribute that checks ConstantExpr for AddrSpaceCast?

jwanggit86 replied:
Found `getConstantAccess()`, which might be what you were referring to. I'm looking into it.

jwanggit86 replied:
> I do not want to duplicate the same function that already exists for the LDS case. Unify these.
> 
> We also should try to avoid doing this walk over all instructions through all constant expressions twice for the two attributes

The latest commit creates a new function in AMDGPUInformationCache that makes use of the existing `getConstantAccess()`, so the code in `getConstantAccess()` is not duplicated. However, the two attributes still walk over all instructions to check on constants separately. To unify these two walks would require the walk to happen earlier and the results (all the constants) be collected. Pls let me know your thoughts on this.

jwanggit86 replied:
@arsenm ping here.

jwanggit86 replied:
@arsenm Any more comments?

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUAttributor.cpp:
```diff
@@ -439,6 +439,26 @@ struct AAAMDAttributesFunction : public AAAMDAttributes {
+
+    SmallPtrSet<const Constant *, 8> VisitedConsts;
+
+    for (Instruction &I : instructions(F)) {
```
Should use checkForAllInstructions instead of manually looking at all instructions

shiltian replied:
I don't think it is generally a good idea to put this logic into `initialize` unless it can directly changes the status of the AA.

jwanggit86 committed:
This commit: (1) use checkForAllInstructions() to check AddrSpaceCast
instructions instead of walking through all instructions (2) move
constHasASCast() to AMDGPUInformationCache to make use the existing
function getConstantAccess().

shiltian commented on llvm/lib/Target/AMDGPU/AMDGPUAttributor.cpp:
```diff
@@ -439,6 +451,36 @@ struct AAAMDAttributesFunction : public AAAMDAttributes {
       return;
     }
+
+    { // FLAT_SCRATCH_INIT
```
This is misuse of `initialize`. This code should go to `updateImpl` of the corresponding AA.

jwanggit86 committed:
Move code from initialize() to needFlatScratch() which is called
in updateImpl().

shiltian commented on llvm/lib/Target/AMDGPU/AMDGPUAttributor.cpp:
```diff
@@ -262,6 +262,18 @@ class AMDGPUInformationCache : public InformationCache {
     return !HasAperture && (Access & ADDR_SPACE_CAST);
   }
 
+  bool constHasASCastFromPrivate(const Constant *C, Function &Fn) {
```
nit: function starts with a verb

shiltian commented on llvm/lib/Target/AMDGPU/AMDGPUAttributor.cpp:
```diff
@@ -683,6 +698,65 @@ struct AAAMDAttributesFunction : public AAAMDAttributes {
+    // Check all AddrSpaceCast instructions. FlatScratchInit is needed if
+    // there is a cast from PRIVATE_ADDRESS.
+    auto AddrSpaceCastNotFromPrivate = [&](Instruction &I) {
+      return static_cast<AddrSpaceCastInst &>(I).getSrcAddressSpace() !=
```
`cast<...>` might be more proper here

jwanggit86 committed:
Some minor code changes.

shiltian commented on llvm/lib/Target/AMDGPU/AMDGPUAttributor.cpp:
```diff
@@ -683,6 +698,65 @@ struct AAAMDAttributesFunction : public AAAMDAttributes {
+    // Check for addrSpaceCast from PRIVATE_ADDRESS in constant expressions
+    auto &InfoCache = static_cast<AMDGPUInformationCache &>(A.getInfoCache());
+
+    Function *F = getAssociatedFunction();
```
Is it possible to merge this into `AddrSpaceCastNotFromPrivate`?

jwanggit86 replied:
`AddrSpaceCastNotFromPrivate` is a predicate used in the checks on the AddrSpaceCast instructions. The for-loop on the other hand checks all the constants. It's not clear to me how these two can be merged.

shiltian replied:
Yeah, that is not ideal, because `A.checkForAllInstructions` uses liveness analysis as well, so it is able to skip dead instructions, while the explicit iteration over instructions doesn't, but indeed here we are lack of an interface that just go through all instructions w/o checking op codes.

shiltian commented on llvm/lib/Target/AMDGPU/AMDGPUAttributor.cpp:
```diff
@@ -683,6 +698,65 @@ struct AAAMDAttributesFunction : public AAAMDAttributes {
+      if (!Callee)
+        return true;
+
+      return Callee->getIntrinsicID() !=
```
IIUC, this attribute should propagate from callee to caller, so you will need to check all function calls, and ask Attributor whether the callee needs it or not.

jwanggit86 replied:
> IIUC, this attribute should propagate from callee to caller, so you will need to check all function calls, and ask Attributor whether the callee needs it or not.

Callees are already checked at the beginning of `updateImpl()` (See the for-loop at lines 475-494). When `needFlatScratchInit()` is reached, only inline asm and intrinsics are left to be further checked.

shiltian reviewed (approved):
LGTM. As a follow-up, it might be a good idea to add an extra interface to the attributor framework to walk through all instructions w/o op code filter, and let the existing one to call it.

jwanggit86 commented:
I plan to submit this PR next week by Nov 26. Pls let me know if you have additional comments.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUAttributor.cpp:
```diff
@@ -683,6 +698,65 @@ struct AAAMDAttributesFunction : public AAAMDAttributes {
+
+    bool UsedAssumedInformation = false;
+    if (!A.checkForAllInstructions(AddrSpaceCastNotFromPrivate, *this,
+                                   {Instruction::AddrSpaceCast},
```
Can't this handle the call case instead of a separate checkForAllCallLikeInstructions?

Alternatively, we should finally add the nonnull flag to addrspacecast
